README.md
Demo pages for deck.gl.

## Development
```
npm install
npm run start
npm run build
```


CHANGELOG.md
<!--
Each version should:
  List its release date in the above format.
  Group changes to describe their impact on the project, as follows:
  Added for new features.
  Changed for changes in existing functionality.
  Deprecated for once-stable features removed in upcoming releases.
  Removed for deprecated features removed in this release.
  Fixed for any bug fixes.
  Security to invite users to upgrade in case of vulnerabilities.
Ref: http://keepachangelog.com/en/0.3.0/
-->

### Change Log
All notable changes to deck.gl will be documented in this file.

#### [Ongoing]
- Tutorials of using deck.gl
- An updated new documentation page(s) to replace the current readme.md
- High-precision (64-bit) floating point calculation support

### Beta Releases

### [3.0.0-beta13] - 2016-09-06
- Remove separate attribute definitions, simplifying layer subclass creation
- Standar

<<<<<<< 2b245b81d86c9934c4382bc112453c85f7b94c78
#### [2.4.7] - 2016-09-06
- Fix issue of disableMercatorProjector not working on Linux

=======
>>>>>>> 2.4.6
#### [2.4.6] - 2016-09-06
- Fix issue where 0 opacity is interpreted as default opacity
- Fix issue with printing of layerName in debug messages crashes

<<<<<<< 2b245b81d86c9934c4382bc112453c85f7b94c78
=======

>>>>>>> 2.4.6
#### [2.4.5] - 2016-08-31
- Fixed picking for the choropleth layer

#### [2.4.4] - 2016-08-17
- Added deckgl-overlay canvas ID and customize style support

#### [2.4.3] - 2016-08-16
- Fix document / add customize style support to the canvas (@contra)

#### [2.4.2] - 2016-08-16
- Added per point radius support for the scatterplot-layer

#### [2.4.1] - 2016-08-15
- Fixed primitive distortion bug for the scatterplot and hexagon-layer

#### [2.4.0] - 2016-08-12
- Added non-LatLng coordinate support for
  - arc-layer
  - choropleth-layer
  - line-layer
  - scatterplot-layer

#### [2.3.0] - 2016-08-06
- Added line-layer support

#### [2.2.5] - 2016-08-02
- Added per point color support for the scatterplot-layer

#### [2.2.4] - 2016-07-13
- Performance improvement

#### [2.2.0] - 2016-07-05
- Added perspective mode, 3D camera support
- Added unit tests
- Tons of refactoring and performance improvement

#### [2.1.0] - 2016-03-30
- Added precompile support
- Added data deep comparison support
- Added better uniform error message support
- Changed to use the new Luma.gl API
- Moved babel-related libraries from devDependence to dependency
- Changed default blending function (ZERO -> ONE_MINUS_SRC_ALPHA)
- Bug in getNumberInstances

#### [2.0.0] - 2016-02-29
- Retina display support
- Performance refactoring
- Switched the underlying rendering framework to [luma.gl](https://github.com/uber/luma.gl)
- Fixed picking on retina/regular display

#### [1.0.0] - 2016-01-06
- Initial commit of the open-source version of deck.gl


example/main.js
// Copyright (c) 2015 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

/* eslint-disable func-style */
/* eslint-disable no-console */
/* global console, process */
import 'babel-polyfill';
import document from 'global/document';
import window from 'global/window';

import React from 'react';
import ReactDOM from 'react-dom';
import {createStore} from 'redux';
import {Provider, connect} from 'react-redux';
import autobind from 'autobind-decorator';

import MapboxGLMap from 'react-map-gl';
import OverlayControl from './overlay-control';

import request from 'd3-request';
import {
  DeckGLOverlay,
  HexagonLayer,
  ChoroplethLayer,
  ScatterplotLayer,
  ArcLayer,
  LineLayer,
  GridLayer
} from '../src';

const BLUE = [0, 0, 255];
const GREEN = [0, 255, 0];

// ---- Default Settings ---- //
/* eslint-disable no-process-env */
const MAPBOX_ACCESS_TOKEN = process.env.MAPBOX_ACCESS_TOKEN ||
  'Set MAPBOX_ACCESS_TOKEN environment variable or put your token here.';

const CHOROPLETHS_FILE = './example/data/sf.zip.geo.json';
const HEXAGONS_FILE = './example/data/hexagons.csv';
const POINTS_FILE = './example/data/sf.bike.parking.csv';

const INITIAL_STATE = {
  mapViewState: {
    latitude: 37.751537058389985,
    longitude: -122.42694203247012,
    zoom: 11.5,
    bearing: 0,
    pitch: 0
  },
  choropleths: null,
  hexagons: null,
  points: null,
  arcs: null,
  arcs2: null,
  lines: null,
  arcStrokeWidth: 1
};

// ---- Action ---- //
function updateMap(mapViewState) {
  return {type: 'UPDATE_MAP', mapViewState};
}

function loadChoropleths(choropleths) {
  return {type: 'LOAD_CHOROPLETHS', choropleths};
}

function loadHexagons(hexagons) {
  return {type: 'LOAD_HEXAGONS', hexagons};
}

function loadPoints(points) {
  return {type: 'LOAD_POINTS', points};
}

// Swaps data props when clicked to trigger WebGLBuffer updates
function swapData() {
  return {type: 'SWAP_DATA'};
}

// ---- Reducer ---- //
function reducer(state = INITIAL_STATE, action) {
  switch (action.type) {
  case 'UPDATE_MAP':
    return {...state, mapViewState: action.mapViewState};
  case 'LOAD_CHOROPLETHS':
    return {...state, choropleths: action.choropleths};
  case 'LOAD_HEXAGONS': {
    const {hexagons} = action;
    const hexData2 = processHexagons(hexagons);
    const hexData = hexData2.slice(hexData2.length / 2);
    return {...state, hexagons, hexData, hexData2};
  }
  case 'LOAD_POINTS': {
    const points = action.points.map(point => {
      const coordString = point.COORDINATES;
      const p0 = coordString.indexOf('(') + 1;
      const p1 = coordString.indexOf(')');
      const coords = coordString.slice(p0, p1).split(',');
      return {
        position: {
          x: Number(coords[1]),
          y: Number(coords[0]),
          z: 0
        },
        color: [88, 9, 124],
        radius: (Math.random() * (15 - 5 + 1) + 5) / 10  // [0.5, 1.5]
      };
    });

    const arcs = pointsToArcs(points);
    const arcs1 = arcs.slice(0, arcs.length / 2);
    const arcs2 = arcs.slice(arcs.length / 2);
    const lines = pointsToLines(points);
    return {...state, points, arcs: arcs1, arcs2, lines};
  }
  case 'SWAP_DATA': {
    state = {
      ...state,
      hexData: state.hexData2,
      hexData2: state.hexData
    };
    return state;
  }
  default:
    return state;
  }
}

// redux states -> react props
function mapStateToProps(state) {
  return {
    mapViewState: state.mapViewState,
    choropleths: state.choropleths,
    hexagons: state.hexagons,
    points: state.points,
    arcs: state.arcs,
    arcs2: state.arcs2,
    lines: state.lines,
    hexData: state.hexData,
    hexData2: state.hexData2
  };
}

// ---- Helpers ---- //

function processHexagons(hexagons) {
  const values = hexagons.map(hexagon => Number(hexagon.value));
  const maxValue = Math.max(...values);

  const data = hexagons.map(hexagon => ({
    centroid: [
      hexagon['centroid.x'],
      hexagon['centroid.y']
    ],
    vertices: [
      [Number(hexagon['v0.x']), Number(hexagon['v0.y'])],
      [Number(hexagon['v1.x']), Number(hexagon['v1.y'])],
      [Number(hexagon['v2.x']), Number(hexagon['v2.y'])],
      [Number(hexagon['v3.x']), Number(hexagon['v3.y'])],
      [Number(hexagon['v4.x']), Number(hexagon['v4.y'])],
      [Number(hexagon['v5.x']), Number(hexagon['v5.y'])]
    ],
    color: [
      Number(hexagon.value) / maxValue * 255,
      Number(hexagon.value) / maxValue * 128,
      Number(hexagon.value) / maxValue * 64
    ],
    elevation: Number(hexagon.value) / maxValue

  }));
  return data;
}

function pointsToArcs(points) {
  return points.map((point, i) => {
    if (i === points.length - 1) {
      return {
        position: {x0: 0, y0: 0, x1: 0, y1: 0},
        color: [35, 81, 128]
      };
    }

    const source = point;
    const target = points[i + 1];

    return {
      position: {
        x0: source.position.x, y0: source.position.y,
        x1: target.position.x, y1: target.position.y
      },
      color: [
        i % 255,
        255 - i % 255,
        Math.floor(i / 255) % 255
      ]
    };
  });
}

function pointsToLines(points) {
  return points.map((point, i) => {
    if (i === points.length - 1) {
      return {
        position: {x0: 0, y0: 0, x1: 0, y1: 0},
        color: [35, 81, 128]
      };
    }

    const source = point;
    const target = points[i + 1];

    return {
      position: {
        x0: source.position.x, y0: source.position.y,
        x1: target.position.x, y1: target.position.y
      },
      color: [0, 0, 255]
    };
  });
}

// ---- View ---- //
class ExampleApp extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedHexagons: [],
      hoverHexagon: null,
      hoverPoint: null,
      hoverArc: null,
      hoverChoropleth: null,
      clickItem: null
    };
  }

  componentWillMount() {
    this._handleResize();
    window.addEventListener('resize', this._handleResize);

    this._loadJsonFile(CHOROPLETHS_FILE, this._handleChoroplethsLoaded);
    this._loadCsvFile(HEXAGONS_FILE, this._handleHexagonsLoaded);
    this._loadCsvFile(POINTS_FILE, this._handlePointsLoaded);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this._handleResize);
  }

  _loadJsonFile(path, onDataLoaded) {
    request.json(path, function loadJson(error, data) {
      if (error) {
        console.error(error);
      }
      onDataLoaded(data);
    });
  }

  _loadCsvFile(path, onDataLoaded) {
    request.csv(path, function loadJson(error, data) {
      if (error) {
        console.error(error);
      }
      onDataLoaded(data);
    });
  }

  @autobind _updateArcStrokeWidth() {
    this.setState({arcStrokeWidth: 1});
  }

  @autobind _handleHexagonsLoaded(data) {
    this.props.dispatch(loadHexagons(data));
  }

  @autobind _handlePointsLoaded(data) {
    this.props.dispatch(loadPoints(data));
  }

  @autobind _handleChoroplethsLoaded(data) {
    this.props.dispatch(loadChoropleths(data));
  }

  @autobind _handleResize() {
    this.setState({width: window.innerWidth, height: window.innerHeight});
  }

  @autobind _handleViewportChanged(mapViewState) {
    if (mapViewState.pitch > 60) {
      mapViewState.pitch = 60;
    }
    this.props.dispatch(updateMap(mapViewState));
  }

  @autobind _handleChoroplethHovered(info) {
    info.type = 'choropleth';
    this.setState({hoverChoropleth: info});
  }

  @autobind _handleChoroplethClicked(info) {
    info.type = 'choropleth';
    this.setState({clickItem: info});
  }

  @autobind _handleHexagonHovered(info) {
    info.type = 'hexagon';

    const {hexData} = this.props;
    let selectedHexagons = [];
    if (info.index >= 0) {
      selectedHexagons = [{
        ...hexData[info.index],
        color: [0, 0, 255]
      }];
    }

    this.setState({
      hoverHexagon: info,
      selectedHexagons
    });
  }

  @autobind _handleHexagonClicked(info) {
    info.type = 'hexagon';
    this.setState({clickItem: info});
    this.props.dispatch(swapData());
  }

  @autobind _handleScatterplotHovered(info) {
    info.type = 'point';
    this.setState({hoverPoint: info});
  }

  @autobind _handleScatterplotClicked(info) {
    info.type = 'point';
    this.setState({clickItem: info});
  }

  @autobind _handleArcHovered(info) {
    info.type = 'arc';
    this.setState({hoverArc: info});
  }

  @autobind _handleArcClicked(info) {
    info.type = 'arc';
    this.setState({clickItem: info});
  }

  @autobind _handleLineHovered(info) {
    info.type = 'line';
    this.setState({hoverLine: info});
  }

  @autobind _handleLineClicked(info) {
    info.type = 'line';
    this.setState({clickItem: info});
  }

  _renderGridLayer() {
    const {mapViewState, points} = this.props;

    return new GridLayer({
      id: 'gridLayer',
      width: window.innerWidth,
      height: window.innerHeight,
      latitude: mapViewState.latitude,
      longitude: mapViewState.longitude,
      zoom: mapViewState.zoom,
      isPickable: false,
      opacity: 0.06,
      data: points
    });
  }

  _renderChoroplethContourLayer() {
    const {mapViewState, choropleths} = this.props;
    return new ChoroplethLayer({
      id: 'choroplethContourLayer',
      width: window.innerWidth,
      height: window.innerHeight,
      latitude: mapViewState.latitude,
      longitude: mapViewState.longitude,
      zoom: mapViewState.zoom,
      data: choropleths,
      opacity: 0.8,
      drawContour: true
    });
  }

  _renderChoroplethLayer() {
    const {mapViewState, choropleths} = this.props;
    return new ChoroplethLayer({
      id: 'choroplethLayer',
      width: window.innerWidth,
      height: window.innerHeight,
      latitude: mapViewState.latitude,
      longitude: mapViewState.longitude,
      zoom: mapViewState.zoom,
      data: choropleths,
      opacity: 0.01,
      isPickable: true,
      onHover: this._handleChoroplethHovered,
      onClick: this._handleChoroplethClicked
    });
  }

  _renderHexagonLayer() {
    const {mapViewState, hexData} = this.props;

    return new HexagonLayer({
      id: 'hexagonLayer',
      width: window.innerWidth,
      height: window.innerHeight,
      latitude: mapViewState.latitude,
      longitude: mapViewState.longitude,
      zoom: mapViewState.zoom,
      data: hexData,
      opacity: 0.5,
      elevation: 200,
      isPickable: true,
      onHover: this._handleHexagonHovered,
      onClick: this._handleHexagonClicked
    });
  }

  _renderHexagonSelectionLayer() {
    const {mapViewState} = this.props;
    const {selectedHexagons} = this.state;

    return new HexagonLayer({
      id: 'hexagonSelectionLayer',
      width: window.innerWidth,
      height: window.innerHeight,
      latitude: mapViewState.latitude,
      longitude: mapViewState.longitude,
      zoom: mapViewState.zoom,
      data: selectedHexagons,
      opacity: 0.1,
      elevation: 200,
      isPickable: false
    });
  }

  _renderScatterplotLayer() {
    const {mapViewState, points} = this.props;

    return new ScatterplotLayer({
      id: 'scatterplotLayer',
      width: window.innerWidth,
      height: window.innerHeight,
      latitude: mapViewState.latitude,
      longitude: mapViewState.longitude,
      zoom: mapViewState.zoom,
      data: points,
      isPickable: true,
      onHover: this._handleScatterplotHovered,
      onClick: this._handleScatterplotClicked
    });
  }

  _renderArcLayer() {
    const {mapViewState, arcs} = this.props;

    return new ArcLayer({
      id: 'arcLayer',
      width: window.innerWidth,
      height: window.innerHeight,
      latitude: mapViewState.latitude,
      longitude: mapViewState.longitude,
      zoom: mapViewState.zoom,
      data: arcs,
      strokeWidth: this.state.arcStrokeWidth || 1,
      color0: BLUE,
      color1: BLUE,
      isPickable: true,
      onHover: this._handleArcHovered,
      onClick: this._handleArcClicked
    });
  }

  _renderArcLayer2() {
    const {mapViewState, arcs2} = this.props;

    return new ArcLayer({
      id: 'arcLayer2',
      width: window.innerWidth,
      height: window.innerHeight,
      latitude: mapViewState.latitude,
      longitude: mapViewState.longitude,
      zoom: mapViewState.zoom,
      data: arcs2,
      strokeWidth: this.state.arcStrokeWidth || 1,
      color0: GREEN,
      color1: GREEN,
      isPickable: true,
      onHover: this._handleArcHovered,
      onClick: this._handleArcClicked
    });
  }

  _renderLineLayer() {
    const {mapViewState, lines} = this.props;

    return new LineLayer({
      id: 'lineLayer',
      width: window.innerWidth,
      height: window.innerHeight,
      latitude: mapViewState.latitude,
      longitude: mapViewState.longitude,
      zoom: mapViewState.zoom,
      data: lines,
      strokeWidth: this.state.lineStrokeWidth || 1,
      isPickable: true,
      onHover: this._handleLineHovered,
      onClick: this._handleLineClicked
    });
  }

  _renderOverlay() {
    const {choropleths, hexagons, points, mapViewState} = this.props;
    const {width, height} = this.state;

    // wait until data is ready before rendering
    if (!choropleths || !points || !hexagons) {
      return [];
    }

    return (
      <DeckGLOverlay
        id="default-deckgl-overlay"
        width={width}
        height={height}
        {...mapViewState}
        layers={[
          // this._renderGridLayer(),
          //  this._renderHexagonLayer(),
          //  this._renderHexagonSelectionLayer(),
          // this._renderArcLayer(),
          // this._renderArcLayer2(),
          // this._renderLineLayer(),
          this._renderScatterplotLayer(),
          // this._renderChoroplethLayer(),
          // this._renderChoroplethContourLayer()
        ]}
        onWebGLInitialized={ this._onWebGLInitialized }
      />
    );
  }

  @autobind _onWebGLInitialized(gl) {
    gl.enable(gl.DEPTH_TEST);
    gl.depthFunc(gl.LEQUAL);
  }

  _renderMap() {
    const {mapViewState} = this.props;
    const {width, height} = this.state;

    return (
      <MapboxGLMap
        mapboxApiAccessToken={MAPBOX_ACCESS_TOKEN}
        width={width}
        height={height}
        perspectiveEnabled
        { ...mapViewState }
        onChangeViewport={this._handleViewportChanged}>
        { this._renderOverlay() }
      </MapboxGLMap>
    );
  }

  render() {
    return (
      <div>
        { this._renderMap() }
        <OverlayControl { ...this.state }/>
      </div>
    );
  }
}

// ---- Main ---- //
const store = createStore(reducer);
const App = connect(mapStateToProps)(ExampleApp);

const container = document.createElement('div');
document.body.appendChild(container);

ReactDOM.render(
  <Provider store={store}>
    <App />
  </Provider>,
  container
);

/* eslint-enable func-style */
/* eslint-enable no-console */
